### PR TITLE
Add tests for BoundedBlockingQueue

### LINKED ISSUE #24733: Tests for BoundedBlockingQueue and AffinityPool
From coverage analysis, we found that both `BoundedBlockingQueue` and `AffinityPool` do not have any tests by themselves, and their behavior is also hardly covered by other tests. 

Would you accept a PR adding some tests for either of these?

### PR DESCRIPTION
**Pull request by txsmith**
Fixes #24733 

This took some more time than I had initially hoped. Turns out testing the whether the queue properly uses locks was harder than expected.

I had to make a small change to the class under test to allow injecting a lock and `Condition`s that we can monitor in the tests.

Let me know what you think!

### CODE REVIEW

**Review by johanandren — APPROVED** [review]
That's a lot of testing infrastructure for a single test, but contained, LGTM I guess

**Comment by johanandren** [comment]
OK TO TEST

**Comment by johanandren** [comment]
@txsmith the test failure is because of incorrect formatting, if you run the test or compile things locally that will trigger the auto formatter, make sure to check in and push the formatted result.

**Comment by txsmith** [comment]
Yes it's indeed quite some infra. I'm not too happy with that myself so if you have any suggestions on how to fix that I'll be happy to follow up.

**Comment by raboof** [comment]
PLS BUILD

File: akka-actor-tests/src/test/scala/akka/util/BoundedBlockingQueueSpec.scala [hunk 1]
```diff
@@ -0,0 +1,818 @@
...
+
+    "not accept a non-empty backing Queue" in {
+      intercept[IllegalArgumentException] {
+        val listQ = new util.LinkedList[String]()
+        listQ.add("Hello")
+        new BoundedBlockingQueue[String](10, listQ)
+      }
+    }
+
+    "accept and empty backing Queue" in {
```
Comment (raboof) [inline akka-actor-tests/src/test/scala/akka/util/BoundedBlockingQueueSpec.scala]: typo 'an'

File: akka-actor-tests/src/test/scala/akka/util/BoundedBlockingQueueSpec.scala [hunk 2]
```diff
@@ -0,0 +1,818 @@
...
+      new BoundedBlockingQueue(10, new util.LinkedList()).size() should equal(0)
+    }
+
+    "get a positive maxCapacity" in {
```
Comment (raboof) [inline akka-actor-tests/src/test/scala/akka/util/BoundedBlockingQueueSpec.scala]: 'get'->'require'?

File: akka-actor-tests/src/test/scala/akka/util/BoundedBlockingQueueSpec.scala [hunk 3]
```diff
@@ -0,0 +1,818 @@
...
+      intercept[IllegalArgumentException] {
+        new BoundedBlockingQueue(0, new util.LinkedList())
+      }
+    }
+
+    "check remainingCapacity for BlockingQueue" in {
```
Comment (raboof) [inline akka-actor-tests/src/test/scala/akka/util/BoundedBlockingQueueSpec.scala]: 'check'->'report'?

File: akka-actor-tests/src/test/scala/akka/util/BoundedBlockingQueueSpec.scala [hunk 4]
```diff
@@ -0,0 +1,818 @@
...
+      intercept[IllegalArgumentException] {
+        new BoundedBlockingQueue(0, new LinkedBlockingQueue(1))
+      }
+      new BoundedBlockingQueue(1, new LinkedBlockingQueue(1)).remainingCapacity() should equal(1)
```
Comment (raboof) [inline akka-actor-tests/src/test/scala/akka/util/BoundedBlockingQueueSpec.scala]: Perhaps also add a check with a number other than '1'?

File: akka-actor-tests/src/test/scala/akka/util/BoundedBlockingQueueSpec.scala [hunk 5]
```diff
@@ -0,0 +1,818 @@
...
+    }
+
+    "block until the backing queue has space" in {
+      val TestContext(queue, events, _, _, _, _) = newBoundedBlockingQueue(1)
+      queue.offer("a")
+
+      val f = Future(queue.put("b"))
+
+      after(10 milliseconds) {
+        queue.take()
```
Comment (raboof) [inline akka-actor-tests/src/test/scala/akka/util/BoundedBlockingQueueSpec.scala]: Perhaps asserting `f.isCompleted should be(false)` before taking would make this test slightly more readable?

**Review by raboof — APPROVED** [review]
Pretty thorough, LGTM, some comments for your consideration.

**Comment by raboof** [comment]
Thanks for your contribution @txsmith!

**Pull request merged**